lib/block_view.go

```diff
@@ -2295,6 +2480,272 @@ func (bav *UtxoView) _disconnectCreatorCoinTransfer(
+	// Sanity-check that it exists.
+	if existingPostEntry == nil || existingPostEntry.isDeleted {
+		return fmt.Errorf("_disconnectUpdateNFT: Post entry for "+
+			"post hash %v doesn't exist; this should never happen",
+			txMeta.NFTPostHash.String())
+	}
+
+	// Revert to the old post entry since we changed NumNFTCopiesForSale.
+	bav._setPostEntryMappings(operationData.PrevPostEntry)
+
+	// Now revert the basic transfer with the remaining operations.
+	return bav._disconnectBasicTransfer(
+		currentTxn, txnHash, utxoOpsForTxn[:operationIndex+1], blockHeight)
+}
+
+func (bav *UtxoView) _disconnectAcceptNFTBid(
+	operationType OperationType, currentTxn *MsgBitCloutTxn, txnHash *BlockHash,
+	utxoOpsForTxn []*UtxoOperation, blockHeight uint32) error {
+
+	// Verify that the last operation is a CreatorCoinTransfer operation
+	if len(utxoOpsForTxn) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: utxoOperations are missing")
+	}
+	operationIndex := len(utxoOpsForTxn) - 1
+	if utxoOpsForTxn[operationIndex].Type != OperationTypeAcceptNFTBid {
+		return fmt.Errorf("_disconnectAcceptNFTBid: Trying to revert "+
+			"OperationTypeAcceptNFTBid but found type %v",
+			utxoOpsForTxn[operationIndex].Type)
+	}
+	txMeta := currentTxn.TxnMeta.(*AcceptNFTBidMetadata)
+	operationData := utxoOpsForTxn[operationIndex]
+	operationIndex--
+
+	// In order to disconnect an accepted bid, we need to do the following:
+	// 	(1) Revert the NFT entry to the previous one with the previous owner.
+	//  (2) Add back all of the bids that were deleted.
+	//  (3) Disconnect payment UTXOs.
+	//  (4) Unspend bidder UTXOs.
+	//  (5) Revert profileEntry to undo royalties added to BitCloutLockedNanos.
+	//  (6) Revert the postEntry since NumNFTCopiesForSale was decremented.
+
+	// (1) Set the old NFT entry.
+	if operationData.PrevNFTEntry == nil || operationData.PrevNFTEntry.isDeleted {
+		return fmt.Errorf("_disconnectAcceptNFTBid: prev NFT entry doesn't exist; " +
+			"this should never happen")
+	}
+
+	prevNFTEntry := operationData.PrevNFTEntry
+	bav._setNFTEntryMappings(prevNFTEntry)
+
+	// Revert the accepted NFT bid history mappings
+	bav._setAcceptNFTBidHistoryMappings(MakeNFTKey(prevNFTEntry.NFTPostHash, prevNFTEntry.SerialNumber), operationData.PrevAcceptedNFTBidEntries)
+
+	// (2) Set the old bids.
+	if operationData.DeletedNFTBidEntries == nil || len(operationData.DeletedNFTBidEntries) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: DeletedNFTBidEntries doesn't exist; " +
+			"this should never happen")
+	}
+
+	for _, nftBid := range operationData.DeletedNFTBidEntries {
+		bav._setNFTBidEntryMappings(nftBid)
+	}
+
+	// (3) Revert payments made from accepting the NFT bids.
+	if operationData.NFTPaymentUtxoKeys == nil || len(operationData.NFTPaymentUtxoKeys) == 0 {
+		return fmt.Errorf("_disconnectAcceptNFTBid: NFTPaymentUtxoKeys was nil; " +
+			"this should never happen")
+	}
+	for _, paymentUtxoKey := range operationData.NFTPaymentUtxoKeys {
```

Comment:
Minor maybe bug alert: I think you need to unadd these in reverse order. If you don't do that then it could result in corruption of the UTXO state.